### PR TITLE
Print best and worst results in a WER report.

### PR DESCRIPTION
**Pull request by DanBmh**
This is also correcting the error that the results with the highest WER instead of the lowest WER are printed.

### CODE REVIEW

**Comment by lissyx** [comment]
@DanBmh Can I ask you why you think it's useful to print the best ones ?

File: evaluate.py [hunk 1]
```diff
@@ -135,20 +134,27 @@ def run_test(init_op, dataset):
...
-            for sample in report_samples:
+            for i, sample in enumerate(report_samples):
                 print('WER: %f, CER: %f, loss: %f' %
                       (sample.wer, sample.cer, sample.loss))
                 print(' - wav: file://%s' % sample.wav_filename)
                 print(' - src: "%s"' % sample.src)
                 print(' - res: "%s"' % sample.res)
                 print('-' * 80)
 
+                if (i == FLAGS.report_count - 1):
```
Comment (lissyx) [inline evaluate.py]: I'm not sure why you do that here?

Comment (reuben) [reply evaluate.py]: To represent a skip in the data, eg:
```
1
2
[...]
9
10
```

**Comment by DanBmh** [comment]
I didnt know for some time that the examples are the worst predictions from the test set. I thought they were chosen randomly and i was wondering why they were always so bad.  So i did ignore the examples after a while. After i found the flag description (which by the way says they are the best results as lower is better) i understood they are the worst results, This is makes more sense than just printing random ones too. 

So i think  printing both the best and the worst will provide new users with an intuitive way to see that we output the worst results.
Another benefit is that if you have a bad performing network you can still see the network learnt something:)

Maybe the best idea is to print median results too, so that you get a more realistic estimate of the prediction quality? Like this:
1
2
[...]
5
6
[...]
9
10

**Comment by victornoriega** [comment]
I think that when you're modeling and trying a lot of configurations, you want to know the flaws of your model but also the best it can show in which cases it can excel. I also think this should be an additional flag to DeepSpeech.py, not something by default.

File: evaluate.py (hunk 1 above)
Comment (lissyx) [reply evaluate.py]: doh

**Comment by lissyx** [comment]
> Maybe the best idea is to print median results too, so that you get a more realistic estimate of the prediction quality? Like this:
> 1
> 2
> [...]
> 5
> 6
> [...]
> 9
> 10

It makes sense, but I have to admit this is not something that ever crossed our mind.

File: evaluate.py [hunk 2]
```diff
@@ -2,27 +2,25 @@
...
 from util.config import Config, initialize_globals
 from util.evaluate_tools import calculate_report
 from util.feeding import create_dataset
 from util.flags import create_flags, FLAGS
-from util.logging import log_error, log_progress, create_progressbar
-from util.helpers import check_ctcdecoder_version; check_ctcdecoder_version()
+from util.helpers import check_ctcdecoder_version
+from util.logging import create_progressbar, log_error, log_progress
+
+check_ctcdecoder_version()
```
Comment (lissyx) [inline evaluate.py]: Please do not change the order and call to `check_ctcdecoder_version()`

File: evaluate.py [hunk 3]
```diff
@@ -132,23 +130,38 @@ def run_test(init_op, dataset):
...
+            # Print test summary
             wer, cer, samples = calculate_report(wav_filenames, ground_truths, predictions, losses)
             mean_loss = np.mean(losses)
+            print('Test on %s - WER: %f, CER: %f, loss: %f' % (dataset, wer, cer, mean_loss))
+            print('-' * 80)
 
-            # Take only the first report_count items
-            report_samples = itertools.islice(samples, FLAGS.report_count)
+            # Take only the first, median and last report_count items
+            best_samples = samples[:FLAGS.report_count]
```
Comment (lissyx) [inline evaluate.py]: I'm unsure about the differences here with `itertools.islice()` wrt your way to do it.

File: evaluate.py [hunk 4]
```diff
@@ -132,23 +130,38 @@ def run_test(init_op, dataset):
...
+            def print_single_sample(sample):
+                print('WER: %f, CER: %f, loss: %f' % (sample.wer, sample.cer, sample.loss))
                 print(' - wav: file://%s' % sample.wav_filename)
                 print(' - src: "%s"' % sample.src)
                 print(' - res: "%s"' % sample.res)
                 print('-' * 80)
 
+            for s in best_samples:
+                print_single_sample(s)
+            print('[...]', '\n' + '-' * 80)
```
Comment (lissyx) [inline evaluate.py]: I'm not sure we really need this

File: evaluate.py [hunk 5]
```diff
@@ -132,23 +130,38 @@ def run_test(init_op, dataset):
...
+            median_right = FLAGS.report_count - median_left
+            median_samples = samples[median_index - median_left:median_index + median_right]
 
-            print('Test on %s - WER: %f, CER: %f, loss: %f' %
-                  (dataset, wer, cer, mean_loss))
-            print('-' * 80)
-            for sample in report_samples:
-                print('WER: %f, CER: %f, loss: %f' %
-                      (sample.wer, sample.cer, sample.loss))
+            def print_single_sample(sample):
```
Comment (lissyx) [inline evaluate.py]: I'd prefer if we had clearly the information in the output that this is `worst` / `median` / `best` samples

File: evaluate.py (hunk 2 above)
Comment (DanBmh) [reply evaluate.py]: Sorry, this was automatic reordering. I undid it again.
But what is this order necessary for?

File: evaluate.py [hunk 6]
```diff
@@ -132,23 +128,38 @@ def run_test(init_op, dataset):
...
             mean_loss = np.mean(losses)
+            print('Test on %s - WER: %f, CER: %f, loss: %f' % (dataset, wer, cer, mean_loss))
+            print('-' * 80)
 
-            # Take only the first report_count items
-            report_samples = itertools.islice(samples, FLAGS.report_count)
+            # Take only the first, median and last report_count items
+            best_samples = samples[:FLAGS.report_count]
+            worst_samples = samples[-FLAGS.report_count:]
+            median_index = int(len(samples) / 1.5)
```
Comment (lissyx) [inline evaluate.py]: That seems like a weird definition of the median ?

File: util/flags.py [hunk 7]
```diff
@@ -118,7 +118,7 @@ def create_flags():
     f.DEFINE_boolean('show_progressbar', True, 'Show progress for training, validation and testing processes. Log level should be > 0.')
 
     f.DEFINE_boolean('log_placement', False, 'whether to log device placement of the operators to the console')
-    f.DEFINE_integer('report_count', 10, 'number of phrases with lowest WER(best matching) to print out during a WER report')
+    f.DEFINE_integer('report_count', 5, 'number of phrases with best WER, median WER and with worst WER to print out during a WER report')
```
Comment (lissyx) [inline util/flags.py]: nit: unclear to the user here whether we will have 5 sentences covering all best, median and worst, or if we have 5 of each

File: evaluate.py (hunk 3 above)
Comment (DanBmh) [reply evaluate.py]: Not much here, but it allows negative indexing for the worst samples.

File: util/evaluate_tools.py [hunk 8]
```diff
@@ -68,10 +67,10 @@ def calculate_report(wav_filenames, labels, decodings, losses):
     # Order the remaining items by their loss (lowest loss on top)
     samples.sort(key=lambda s: s.loss)
 
-    # Then order by descending WER/CER
+    # Then order by ascending WER/CER
```
Comment (lissyx) [inline util/evaluate_tools.py]: Why do you change that ?

File: util/evaluate_tools.py [hunk 9]
```diff
@@ -5,7 +5,6 @@
 from multiprocessing.dummy import Pool
 
 from attrdict import AttrDict
-
```
Comment (lissyx) [inline util/evaluate_tools.py]: nit: don't remove empty lines

File: evaluate.py (hunk 2 above)
Comment (lissyx) [reply evaluate.py]: It's the same in other parts of the code, and it makes more sense to tie them together.

**Comment by lissyx** [comment]
@DanBmh You might need to factorize / apply that to `evaluate.py` as well as `evaluate_tflite.py`.

File: util/evaluate_tools.py (hunk 8 above)
Comment (DanBmh) [reply util/evaluate_tools.py]: I think the sorting should be in a way that you see the worst WER and the worst loss, instead of the worst WER with the best lost. Did not make sense for me...

This could also be done by sorting loss in reversed order but then i need to reverse it a second time for printing

**Comment by DanBmh** [comment]
evaluate_tflite.py does not print any samples. Shall i add it?

File: evaluate.py (hunk 6 above)
Comment (DanBmh) [reply evaluate.py]: Your right^^
I did test something else with that...

File: evaluate.py (hunk 5 above)
Comment (DanBmh) [reply evaluate.py]: Like this:

-----
Examples with best WER score:
----
... some examples ...
-----
Examples with median WER score:
----
... some examples ...
-----
Examples with worst WER score:
----
... some examples ...
----

?

File: util/evaluate_tools.py (hunk 8 above)
Comment (reuben) [reply util/evaluate_tools.py]: The worst WER with the best loss is to identify systemic issues, where the acoustic model is confident, yet the result is completely off the mark. This can point to transcription errors and stuff like that. Please don't change it.

**Comment by reuben** [comment]
Thanks for the PR!

**Comment by lissyx** [comment]
> evaluate_tflite.py does not print any samples. Shall i add it?

I'd say no then.

File: evaluate.py (hunk 5 above)
Comment (lissyx) [reply evaluate.py]: No need to make the output too heavy in text, but that's the idea, yeah. It could just be worded as `Worst WER:`, `Median WER:` and `Best WER:`

File: evaluate.py (hunk 3 above)
Comment (lissyx) [reply evaluate.py]: I'd worry about performances differences, that's all.

File: evaluate_tflite.py [hunk 10]
```diff
@@ -98,11 +98,14 @@ def main(args, _):
...
+    # Print test summary
     wer, cer, samples = calculate_report(wav_filenames, ground_truths, predictions, losses)
     mean_loss = np.mean(losses)
+    print('Test - WER: %f, CER: %f, loss: %f' % (wer, cer, mean_loss))
+    print('-' * 80)
 
-    print('Test - WER: %f, CER: %f, loss: %f' %
-          (wer, cer, mean_loss))
+    # Print some examples
+    print_report(samples)
```
Comment (lissyx) [inline evaluate_tflite.py]: Wouldn't it make sense to factorize a bit more and also include the `print()` statement above into `print_report` ?

File: util/evaluate_tools.py [hunk 11]
```diff
@@ -65,13 +65,43 @@ def calculate_report(wav_filenames, labels, decodings, losses):
...
+    def print_single_sample(sample):
+        print('WER: %f, CER: %f, loss: %f' % (sample.wer, sample.cer, sample.loss))
+        print(' - wav: file://%s' % sample.wav_filename)
+        print(' - src: "%s"' % sample.src)
+        print(' - res: "%s"' % sample.res)
+        print('-' * 80)
+
+    for s in best_samples:
+        print_single_sample(s)
+    print('[...]', '\n' + '-' * 80)
```
Comment (lissyx) [inline util/evaluate_tools.py]: Still not sold on the `[...]` thing :)

File: util/flags.py [hunk 12]
```diff
@@ -118,7 +118,7 @@ def create_flags():
     f.DEFINE_boolean('show_progressbar', True, 'Show progress for training, validation and testing processes. Log level should be > 0.')
 
     f.DEFINE_boolean('log_placement', False, 'whether to log device placement of the operators to the console')
-    f.DEFINE_integer('report_count', 10, 'number of phrases with lowest WER(best matching) to print out during a WER report')
+    f.DEFINE_integer('report_count', 5, 'number of phrases for each of best WER, median WER and worst WER to print out during a WER report')
```
Comment (lissyx) [inline util/flags.py]: :+1:

File: evaluate_tflite.py (hunk 10 above)
Comment (DanBmh) [reply evaluate_tflite.py]: I thought about this but they are different. evaluate.py also prints the datset path, evaluate_tflite.py doesnt.

File: util/evaluate_tools.py (hunk 11 above)
Comment (DanBmh) [reply util/evaluate_tools.py]: Is my suggestion above better?

Comment (DanBmh) [reply util/evaluate_tools.py]: https://github.com/mozilla/DeepSpeech/pull/2724#discussion_r375764775

File: evaluate.py (hunk 3 above)
Comment (DanBmh) [reply evaluate.py]: According to this question: https://stackoverflow.com/q/2738096
it should be faster than islice(). i think its also more readable.

Comment (lissyx) [reply evaluate.py]: Yeah, but it's not unlikely that we had that for a reason, and I can't recall why. maybe @reuben remembers?

File: evaluate_tflite.py (hunk 10 above)
Comment (lissyx) [reply evaluate_tflite.py]: I don't think it would be a big deal ; it's likely a side-effect that in the past other code was behaving like `evaluate_tflite.py` and over time because we kept same duplicated code, they diverged.

Comment (DanBmh) [reply evaluate_tflite.py]: So move this too?
And leave out the dataset name or make it an optional param?

Comment (lissyx) [reply evaluate_tflite.py]: > So move this too?
I'd say yes

> And leave out the dataset name or make it an optional param?
Sorry, I'm confused to what you refer here.

Comment (DanBmh) [reply evaluate_tflite.py]: tflite: print('Test - WER: %f, CER: %f, loss: %f' % (wer, cer, mean_loss))
eval: print('Test on %s - WER: %f, CER: %f, loss: %f' % (dataset, wer, cer, mean_loss))

Comment (lissyx) [reply evaluate_tflite.py]: Just include the dataset. There's no good reason not to show it.

Comment (DanBmh) [reply evaluate_tflite.py]: there is no dataset in tflite. only a csv file the user sets as argument. i think that was the reason not to show it.
But we can use the csv path instead of the dataset name.

Comment (lissyx) [reply evaluate_tflite.py]: > there is no dataset in tflite. only a csv file the user sets as argument. i think that was the reason not to show it.
> But we can use the csv path instead of the dataset name.

fine by me :)

**Comment by lissyx** [comment]
I've triggered some PR to have TaskCluster running on that.

**Comment by lissyx** [comment]
Thanks @DanBmh !
